File: Problems/Algorithms/797. All Paths from Source to Target/all_paths.go
func allPathsSourceTarget(graph [][]int) [][]int {
    results := make([][]int, 0)
    path := make([]int, 0)

    path = append(path, 0)
    aux(graph, &results, path, 0, len(graph)-1)
    return results
}

func aux(graph [][]int, results *[][]int, path []int, src int, target int) {
    if src == target {
        *results = append(*results, append([]int{}, path...))
        return
    }

    if len(graph[src]) == 0 {
        return
    }

    for _, neighbor := range graph[src] {
        path = append(path, neighbor)
        aux(graph, results, path, neighbor, target)
        path = path[:len(path)-1]
    }
}

File: Problems/Algorithms/797. All Paths from Source to Target/all_paths.rs
impl Solution {
    pub fn all_paths_source_target(graph: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let mut results = vec![];
        let mut path = vec![0];
        
        aux(&graph, &mut results, &mut path, 0, graph.len()-1);
        results
    }
}

pub fn aux(graph: &Vec<Vec<i32>>, results: &mut Vec<Vec<i32>>, path: &mut Vec<i32>, src: i32, target: usize) {
    if src as usize == target {
        results.push(path.clone());
        return
    }
        
    if graph[src as usize].len() == 0 {
        return
    }
        
    for neighbor in graph[src as usize].iter().copied() {
        path.push(neighbor);
        aux(graph, results, path, neighbor, target);
        path.pop();
    }
}

